contracts/ring.md

```diff
@@ -0,0 +1,340 @@
+  state. However, the state of an inherited contract cannot be
+  accessed other than through internal functions defined by that
+  contract.
+- Open question: should we allow overriding defined functions? I would
+  suggest no, but there might be compelling use cases that I'm
+  missing.
+
+### Mutable state
+
+Ring does not have arbitrary mutable state, but only a limited form of
+state associated with each contract instance.
+
+- Each contract defines a state type encapsulating its mutable state.
+- The value of the state is accessible from inside the contract
+  through an implicitly bound variable state.
+- State updates are performed by calling a function put : state =>
+  unit (possibly with a better name).
+- Aside from the put function (and similar functions for transactions
+  and events), the contract language is purely functional.
+- Functions modifying the state need to be annotated with the stateful keyword.
+
+To make it convenient to update parts of a deeply nested state Ring
+provides special special syntax for map/record updates.  Open
+question: we likely want to make it possible to have immutable state
+(parameters). Keep separate from mutable state or annotate certain
+fields as immutable?
+
+### Types
+Ring has the following types:
+
+| Type    | Description                     | Example
+| ------- | ------------------------------- | -------:
+| uint    | A 256 bit integer               | ```42```
+| int     | A 256 bit 2-complement integer  | ```-1```
+| address | A 256 bit number given as a hex | ```ff00```
+| bool    | A Boolean                       | ```true```
+| string  | An array of bytes               | ```“Foo”```
+| list    | A homogeneous immutable singly linked list. | ```[1, 2, 3]```
+| tuple   | An ordered heterogeneous array   | ```(42, “Foo”, true)```
+| record  | An immutable key value store with fixed key names and typed values | ``` type balance = { owner: address, value: uint } ```
+| map     | An immutable key value store with dynamic mapping of keys of one type to values of one type | ```type accounts = map(string, address)```
+| state   | A record of blockstate key, value pairs  |
+| transactions | An append only list of blockchain transactions |
+| events   | An append only list of blockchain events (or log entries) |
+
+#### Arrays
+Ring might not have arrays… Ring might have Maps.
+
+####  Variants
+Depending on how we write the compiler, if they are cheap to include
+some like to use variant types, while others might not want them.
+
+####  Refs
+ Ring has no refs.
+
+####  Object
+ Ring will not have objects. (only contracts) 
+
+### Pattern matching
+Pattern matching is probably outside the first iteration of Ring, but
+we definitely want it in the final language.
+
+### Builtins
+#### Events
+Ring has events which logs a structured message to the contract log in
+the resulting blockchain transaction.
+
+```
+event(name: string, value: any)
+```
```

Comment:
We probably want event types to be declared.

The simplest option would be that a contract can declare a single `event` type (which can be a variant type) and the event function takes an argument of this type.